src/logger/logger.ts

```diff
@@ -0,0 +1,668 @@
+  /**
+   * The stream to write to.  Mutually exclusive with `path`.
+   */
+  stream?: Writable;
+  /**
+   * The name of the stream.
+   */
+  name?: string;
+  /**
+   * A log file path to write to.  Mutually exclusive with `stream`.
+   */
+  path?: string;
+}
+
+/**
+ * Any numeric `Logger` level.
+ */
+export type LoggerLevelValue = LoggerLevel | number;
+
+/**
+ * An object
+ */
+export type Fields = Record<string, unknown>;
+
+/**
+ * All possible field value types.
+ *
+ * @deprecated.  Fields can be anything, including another object
+ */
+export type FieldValue = string | number | boolean;
+
+/**
+ * Log line interface
+ */
+export interface LogLine {
+  name: string;
+  hostname: string;
+  pid: string;
+  log: string;
+  level: number;
+  msg: string;
+  time: string;
+  v: number;
+}
+
+/**
+ * A logging abstraction powered by {@link https://github.com/pinojs/pino | Pino} that provides both a default
+ * logger configuration that will log to the default path, and a way to create custom loggers based on the same foundation.
+ *
+ * ```
+ * // Gets the root sfdx logger
+ * const logger = await Logger.root();
+ *
+ * // Creates a child logger of the root sfdx logger with custom fields applied
+ * const childLogger = await Logger.child('myRootChild', {tag: 'value'});
+ *
+ * // Creates a custom logger unaffiliated with the root logger
+ * const myCustomLogger = new Logger('myCustomLogger');
+ *
+ * // Creates a child of a custom logger unaffiliated with the root logger with custom fields applied
+ * const myCustomChildLogger = myCustomLogger.child('myCustomChild', {tag: 'value'});
+ *
+ * // get a raw pino logger from the root instance of Logger
+ * // you can use these to avoid constructing another Logger wrapper class and to get better type support
+ * const logger = Logger.getRawRootLogger().child({name: 'foo', otherProp: 'bar'});
+ * logger.info({some: 'stuff'}, 'a message');
+ *
+ *
+ * // get a raw pino logger from the current instance
+ * const childLogger = await Logger.child('myRootChild', {tag: 'value'});
+ * const logger = childLogger.getRawLogger();
+ * ```
+ *
+ * **See** https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_dev_cli_log_messages.htm
+ */
+export class Logger {
+  /**
+   * The name of the root sfdx `Logger`.
+   */
+  public static readonly ROOT_NAME = 'sf';
+
+  /**
+   * The default `LoggerLevel` when constructing new `Logger` instances.
+   */
+  public static readonly DEFAULT_LEVEL = LoggerLevel.WARN;
+
+  /**
+   * A list of all lower case `LoggerLevel` names.
+   *
+   * **See** {@link LoggerLevel}
+   */
+  public static readonly LEVEL_NAMES = Object.values(LoggerLevel)
+    .filter(isString)
+    .map((v: string) => v.toLowerCase());
+
+  // The sfdx root logger singleton
+  private static rootLogger?: Logger;
+
+  /**
+   * @deprecated.  Has no effect
+   */
+
+  public readonly logRotationCount = new Env().getNumber('SF_LOG_ROTATION_COUNT') ?? 2;
```

Comment:
We could support this by comparing it against `SF_LOG_ROTATION_PERIOD` in the `cleanup` method. Right now it does not look like users are able to configure keeping logs older than 7 days (ignoring the cleanup odds)